pkg/scalers/graphite_scaler.go

```diff
@@ -0,0 +1,224 @@
+	}
+
+	if val, ok := config.TriggerMetadata[grapThreshold]; ok && val != "" {
+		t, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing %s: %s", grapThreshold, err)
+		}
+
+		meta.threshold = t
+	}
+
+	authModes, ok := config.TriggerMetadata["authModes"]
+	// no authMode specified
+	if !ok {
+		return &meta, nil
+	}
+
+	authTypes := strings.Split(authModes, ",")
+	for _, t := range authTypes {
+		authType := authentication.Type(strings.TrimSpace(t))
+		switch authType {
+		case authentication.BasicAuthType:
+			if len(config.AuthParams["username"]) == 0 {
+				return nil, errors.New("no username given")
+			}
+
+			meta.username = config.AuthParams["username"]
+			// password is optional. For convenience, many application implement basic auth with
+			// username as apikey and password as empty
+			meta.password = config.AuthParams["password"]
+			meta.enableBasicAuth = true
+		default:
+			return nil, fmt.Errorf("err incorrect value for authMode is given: %s", t)
+		}
+	}
+
+	return &meta, nil
+}
+
+func (s *graphiteScaler) IsActive(ctx context.Context) (bool, error) {
+	val, err := s.ExecuteGrapQuery()
+	if err != nil {
+		graphiteLog.Error(err, "error executing graphite query")
+		return false, err
+	}
+
+	return val > 0, nil
+}
+
+func (s *graphiteScaler) Close() error {
+	return nil
+}
+
+func (s *graphiteScaler) GetMetricSpecForScaling() []v2beta2.MetricSpec {
+	targetMetricValue := resource.NewQuantity(int64(s.metadata.threshold), resource.DecimalSI)
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: kedautil.NormalizeString(fmt.Sprintf("%s-%s-%s", "graphite", s.metadata.serverAddress, s.metadata.metricName)),
+		},
+		Target: v2beta2.MetricTarget{
+			Type:         v2beta2.AverageValueMetricType,
+			AverageValue: targetMetricValue,
+		},
+	}
+	metricSpec := v2beta2.MetricSpec{
+		External: externalMetric, Type: externalMetricType,
+	}
+	return []v2beta2.MetricSpec{metricSpec}
+}
+
+func (s *graphiteScaler) ExecuteGrapQuery() (float64, error) {
+	client := &http.Client{}
+	queryEscaped := url_pkg.QueryEscape(s.metadata.query)
+	url := fmt.Sprintf("%s/render?from=%s&target=%s&format=json", s.metadata.serverAddress, s.metadata.from, queryEscaped)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return -1, err
+	}
+	if s.metadata.enableBasicAuth {
+		req.SetBasicAuth(s.metadata.username, s.metadata.password)
+	}
+	r, err := client.Do(req)
+	if err != nil {
+		return -1, err
+	}
+
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return -1, err
+	}
+	r.Body.Close()
+
+	var result grapQueryResult
+	err = json.Unmarshal(b, &result)
+	if err != nil {
+		return -1, err
+	}
+
+	if len(result) == 0 {
+		return 0, nil
+	} else if len(result) > 1 {
+		return -1, fmt.Errorf("graphite query %s returned multiple series", s.metadata.query)
+	}
+
+	var v float64 = -1
+	for _, datapoints := range result[0].Datapoints {
```

Comment:
I think this is the source of the timestamp. A Graphite datapoint is an array with two values. The first is the actual metric value and the second is the unix epoch timestamp. (See https://graphite-api.readthedocs.io/en/latest/api.html#json for an example.) Given that, this code will return the unix epoch timestamp if and when it is larger than the metric value. 

As another point of reference, here is how a colleague and I modeled the Graphite response when we added Graphite support to Flagger: https://github.com/fluxcd/flagger/blob/main/pkg/metrics/providers/graphite.go#L36-L100.

Reply:
Thanks for the examples. I'm still not convince that the logic as written is somehow returning a timestamp, because first just the logic works in my head, but second I was seeing the correct value being returned in my debugger.    

After taking a quick look at prometheus's scaler again, prometheus only returns the latest value of the query. If that's good enough for prometheus, that's what I'll do here as well rather than this scheme of finding the largest value.

I'll take another stab at this tonight. Thanks for the pointers!

Reply:
My bad -- I misread this loop. I agree it should be returning the max value of the first item in each datapoint, which should be the value. 

I'd agree that being consistent with the prometheus implementation is the way to go.